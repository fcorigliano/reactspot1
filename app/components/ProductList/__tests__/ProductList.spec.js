const React = require('react');
const ProductList = require('..');
const { render, screen, act, fireEvent } = require('@testing-library/react');
const mockProducts = require('./sample.json');

/**
 * NOTA: A medida que avancemos con las ejercitaciones, es importante que actualices
 * la snapshot para que el primer test siga pasando. Esto lo podés hacer borrando la
 * carpeta /__tests__s/snapshots de este componente o, si estás corriendo los tests con el 
 * comando npm run test:unit:watch <nombreComponente>, simplemente apretar la tecla
 * `u` cuando terminen de correr los tests para "updatear" la snapshot.
 */

describe('Ejercicio 1 - El componente ProductList', () => {
    let component;
    const i18n = { gettext: text => text };

    beforeEach(() => {
        component = render(<ProductList i18n={i18n} products={mockProducts}/>);
    });

    test('1) Renderiza', () => {
        const { asFragment } = component;
        expect(asFragment()).toMatchSnapshot();
    });

    test('2) Renderiza una lista de productos recibida por props', () => {
        const productList = screen.getAllByRole('listitem');
        const motorola = screen.getByText(/motorola/i);
        expect(productList).toHaveLength(6);
        expect(motorola).toBeInTheDocument();
    });
});

xdescribe('Ejercicio 2 - El componente ProductList', () => {
    let component;
    const i18n = { gettext: text => text };

    beforeEach(() => {
        component = render(<ProductList i18n={i18n} products={[]}/>);
    });

    test('3) Renderiza el string `No se encontraron productos.` si la API no devuelve productos', () => {
        const msg = screen.getByText(/no se encontraron productos/i);
        expect(msg).toBeInTheDocument();
    });
});

xdescribe('Ejercicio 3 - El componente ProductList', () => {
    let component;
    const i18n = { gettext: text => text };
    const mockConsole = jest.spyOn(console, "log");

    beforeEach(() => {
        component = render(<ProductList i18n={i18n} products={mockProducts}/>);
    });

    test('4) Renderiza un array de productos, cada uno con un botón para agregar a favoritos', () => {
        const button = screen.getAllByRole('button')[0];
        expect(button).toBeInTheDocument();
        expect(button.innerHTML).toMatch(/agregar a favoritos/i);
    });

    test('5) Permite agregar un producto de la lista al array de favoritos', () => {
        act(() => {
            const button = screen.getAllByRole('button')[0];
            fireEvent.click(button);
        });
        
        expect(mockConsole).toHaveBeenLastCalledWith(["MLA1131686227"]);
    });

    test('6) Luego de haber sido agregado, renderiza un botón que permite quitar el producto del array', () => {
        act(() => {
            const addButton = screen.getAllByRole('button')[0];
            fireEvent.click(addButton);
        });

        const deleteButton = screen.getAllByRole('button')[0];
        expect(deleteButton.innerHTML).toMatch(/quitar de favoritos/i);
    }); 

    test('7) El botón efectivamente elimina el elemento del array', () => {
        act(() => {
            const addButton = screen.getAllByRole('button')[0];
            fireEvent.click(addButton);
        });

        act(() => {
            const deleteButton = screen.getAllByRole('button')[0];
            fireEvent.click(deleteButton);
        });
        
        expect(mockConsole).toHaveBeenLastCalledWith([]);
    });
});