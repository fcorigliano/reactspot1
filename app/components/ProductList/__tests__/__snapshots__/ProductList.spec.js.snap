// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ejercicio 1 - El componente ProductList 1) Renderiza 1`] = `
<DocumentFragment>
  <ol>
    <li
      class="card"
    >
      <article>
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy A03 128 Gb Negro 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_737913-MLA49433851234_032022-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy A03 128 Gb Negro 4 Gb Ram
          </h4>
          <h3
            class="price"
          >
            $38999
          </h3>
        </div>
        <button>
          Agregar a favoritos
        </button>
      </article>
    </li>
    <li
      class="card"
    >
      <article>
        <figure
          class="img"
        >
          <img
            alt="Motorola Edge 20 Lite 128 Gb Gris 6 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_833619-MLA48233725917_112021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Motorola Edge 20 Lite 128 Gb Gris 6 Gb Ram
          </h4>
          <h3
            class="price"
          >
            $69999
          </h3>
        </div>
        <button>
          Agregar a favoritos
        </button>
      </article>
    </li>
    <li
      class="card"
    >
      <article>
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy S20 Fe 128 Gb Cloud Navy 6 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_995741-MLA47861036467_102021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy S20 Fe 128 Gb Cloud Navy 6 Gb Ram
          </h4>
          <h3
            class="price"
          >
            $112999
          </h3>
        </div>
        <button>
          Agregar a favoritos
        </button>
      </article>
    </li>
    <li
      class="card"
    >
      <article>
        <figure
          class="img"
        >
          <img
            alt=" Moto G22 128 Gb Cosmic Black 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_854793-MLA50262132804_062022-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
             Moto G22 128 Gb Cosmic Black 4 Gb Ram
          </h4>
          <h3
            class="price"
          >
            $53000
          </h3>
        </div>
        <button>
          Agregar a favoritos
        </button>
      </article>
    </li>
    <li
      class="card"
    >
      <article>
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy A22 128 Gb Black 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_774406-MLA47777566692_102021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy A22 128 Gb Black 4 Gb Ram
          </h4>
          <h3
            class="price"
          >
            $92899
          </h3>
        </div>
        <button>
          Agregar a favoritos
        </button>
      </article>
    </li>
    <li
      class="card"
    >
      <article>
        <figure
          class="img"
        >
          <img
            alt="Samsung Galaxy A32 128 Gb Awesome Black 4 Gb Ram"
            class=""
            decoding="async"
            src="http://http2.mlstatic.com/D_922219-MLA48428094384_122021-I.jpg"
          />
        </figure>
        <div
          class="info-products"
        >
          <h4
            class="title-product"
          >
            Samsung Galaxy A32 128 Gb Awesome Black 4 Gb Ram
          </h4>
          <h3
            class="price"
          >
            $63999
          </h3>
        </div>
        <button>
          Agregar a favoritos
        </button>
      </article>
    </li>
  </ol>
</DocumentFragment>
`;
